### ISSUE TITLE
优化代码审查

### ISSUE THREAD

**Issue by twhice** [Bylx666/key-lang#12]
### 对了，给你看看我很可能不是VSCode的VSCode很可能删掉那行遮羞布的输出

![image](https://github.com/Bylx666/key-lang/assets/97959381/2a796d92-9f27-4da8-8bc5-bdf55909371c)

你知道什么是clippy吗？[链接](https://github.com/rust-lang/rust-clippy)挂给你，好好了解一下

**Comment by Bylx666** [comment]
一丝不挂地表达出自己的优越, 我认可你的勇气. 这说到底是我给自己写的东西, 一开始就没抱多大期望有很多人给我改代码. 

说到底你所谓的"优化"或许只是你的一厢情愿罢了. 用Rust贴近汇编的思维方式并不可耻, 懒得写生命周期, "滥用"static mut同样是我自己的选择. 

我不会关闭该issue, 我也不反对你说的问题. 这一切是我咎由自取, 我会将此issue挂起, 在有所谓的"优化"的准备时拿来参考. 

十分感谢你的直言不讳.

**Comment by limuy2022** [comment]
中肯的评价（

**Comment by yiyueawa** [comment]
> 一丝不挂地表达出自己的优越, 我认可你的勇气. 这说到底是我给自己写的东西, 一开始就没抱多大期望有很多人给我改代码. 
> 
> 说到底你所谓的"优化"或许只是你的一厢情愿罢了. 用Rust贴近汇编的思维方式并不可耻, 懒得写生命周期, "滥用"static mut同样是我自己的选择. 
> 
> 我不会关闭该issue, 我也不反对你说的问题. 这一切是我咎由自取, 我会将此issue挂起, 在有所谓的"优化"的准备时拿来参考. 
> 
> 十分感谢你的直言不讳. 

我的优越是您赋予的，对于您认可我的"勇气"这件事我感到非常诧异，难道对你这种完全不尊重其他人，没有进行格式化的胡乱排版的代码，对这种没有通过clippy，不符合rust代码规范的代码提出意见也是"勇气"？我倒是很羡慕你可以沉浸在你自己的世界里，也很羡慕你可以享受在外行的崇拜，和自我满意中高潮的快乐

你开心就好，我也没指望你这个狂傲的人可以听进来自社区的意见，更是从来不对你可能会去了解代码规范，去了解什么叫代码审查，去了解如何尊重别人抱有任何希望。你满意就行了，虽然你的代码不符合社区的规范(不通过clippy lint，胡乱unsafe且没有通过miri的测试，没有格式化……不对，和你扯这些干什么呢？你都不知道clippy，lint，miri这些东西是什么呢)

希望你能够继续高潮下去，言尽于此

**Comment by Bylx666** [comment]
你都给我看笑了, 希望你在1年之后回来看你的的发言不会脸红

**Comment by TabNahida** [comment]
Having a fight? Let me see what is happening. Interesting, isn't it. 
From my perspective, Rust is a programming language for safety. If you do not follow the programming guidelines, why don't you use C++. C++ is good programming language without strict ``rules``, especially from other C++ programmers. Yet, it is still a good project. Many improvements should be done before upload a video on internet. And what I mean is English... It is my first time that I think its better to understand the meaning with my translator on.

**Comment by TabNahida** [comment]
> 你都给我看笑了, 希望你在1年之后回来看你的的发言不会脸红

And I think she would not. But still you should check your grammar, in both, please.

**Comment by yiyueawa** [comment]
> 你都给我看笑了, 希望你在1年之后回来看你的的发言不会脸红

你都给我看笑了, 希望你在1年之后回来看你的的发言和代码不会脸红🤣🤣🤣

**Comment by yiyueawa** [comment]
> 你都给我看笑了, 希望你在1年之后回来看你的的发言不会脸红

You are right, but your code is still a big shit without any improvement for two days. non of issus i report you fixed because you have never spent any time to know what i said, about lint, clippy, ub, miri… You are beliving in yourself and take no adive from community. My code is as bad as this two years ago, but it had improved by taking with other and learn from blogs… If i see my code what is written by me two year ago, i will also say that IT'S SHIT.Feeling ashamed for my taking？lol，you will if you improve yourself in the feature, but i think it will not happen.

**Comment by Jedsek** [comment]
why you close this issue?

**Comment by Jedsek** [comment]
You have said you would never close this issue && what he said is right,so why you close this issue?

**Comment by Jedsek** [comment]
Please reopen this issue

**Comment by Bylx666** [comment]
把话说干净就好了, 我只是看不惯他的语气而已

**Comment by Jedsek** [comment]
> 把话说干净就好了, 我只是看不惯他的语气而已

yeah，It's great to take advice from others

**Comment by Bylx666** [comment]
看得出来你的语气有所悔改. 我会再次关闭该issue, 创建一个新的`clippy pass`的issue. 该次争吵希望你能认识到自己语气上的问题, 我宣告该次争论到此为止.

**Comment by limuy2022** [comment]
你不应该使用高高在上的语气，他的语气就算再不好也是确确实实的帮你改正错误，他甚至为你提了个pr，而你只是一直在嘴硬，甚至因此拒绝改正自己的错误，适当听取他人建议才有利于提升自己

**Comment by zhleyai** [comment]
Hey, kid, you can take this as a hobby, but you can’t take it as a career, otherwise you will starve to death. My team members and I conducted a preliminary review of the code. This is completely a hobby, and it can’t even be said to be a project. It is a simple REDME, without a future list, and does not meet the requirements of [Software design pattern,](https://en.wikipedia.org/wiki/Software_design_pattern) and [Compilers: Principles.](https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools) I hope you can get admission to the university. Notices can improve one's own quality and knowledge

**Comment by WenSimEHRP** [comment]
> 看得出来你的语气有所悔改. 我会再次关闭该issue, 创建一个新的`clippy pass`的issue. 该次争吵希望你能认识到自己语气上的问题, 我宣告该次争论到此为止.

From your attitude I can tell that you just want to show off. You've kept on rejecting others' kind improvements and responded like you don't give a shit.

It is you who should change the tone. They have been trying to solve the problem so much, even too much. *Anyone* would not like your *educated* response.

If you don't want others to change your code just write it in the readme file, or better, make the repo private so that nobody could see it. What exactly are you trying to do?

---

Also, some minor improvements to your sentence. I'm pretty sure that this sentence won't pass the [HSK](https://en.wikipedia.org/wiki/Hanyu_Shuiping_Kaoshi) check.
```diff
- 看得出来你的语气有所悔改. 我会再次关闭该issue, 创建一个新的`clippy pass`的issue. 该次争吵希望你能认识到自己语气上的问题, 我宣告该次争论到此为止.
+ 看得出来你有所悔改。我会再次关闭该issue并创建一个新的`clippy pass` issue。希望你能认识到自己在该次争吵中语气上的问题，我宣告该次争论到此为止。
```

**Comment by zhleyai** [comment]
I don't understand what you said about repentance, but it is certainly the best thing if you can accept relevant suggestions.




 ----------回复的邮件信息----------
 Jeremy ***@***.***&gt; 于2024年5月6日(周一) 上午9:42写道:

  

  
看得出来你的语气有所悔改. 我会再次关闭该issue, 创建一个新的clippy pass的issue. 该次争吵希望你能认识到自己语气上的问题, 我宣告该次争论到此为止.
  
From your attitude I can tell that you just want to show off. You've kept on rejecting others' kind improvements and responded like you don't give a shit.
 
It is you who should change the tone. They have been trying to solve the problem so much, even too much. Anyone would not like your educated response.
 
If you don't want others to change your code just write it in the readme file, or better, make the repo private so that nobody could see it. What exactly are you trying to do?
 
 
Also, some minor improvements to your sentence. I'm pretty sure that this sentence won't pass the HSK check.
 - 看得出来你的语气有所悔改. 我会再次关闭该issue, 创建一个新的`clippy pass`的issue. 该次争吵希望你能认识到自己语气上的问题, 我宣告该次争论到此为止. + 看得出来你的语气有所悔改。我会再次关闭该issue，创建一个新的`clippy pass`的issue。该次争吵希望你能认识到自己语气上的问题，我宣告该次争论到此为止。

**Comment by TabNahida** [comment]
> > 看得出来你的语气有所悔改. 我会再次关闭该issue, 创建一个新的`clippy pass`的issue. 该次争吵希望你能认识到自己语气上的问题, 我宣告该次争论到此为止.
> 
> From your attitude I can tell that you just want to show off. You've kept on rejecting others' kind improvements and responded like you don't give a shit.
> 
> It is you who should change the tone. They have been trying to solve the problem so much, even too much. _Anyone_ would not like your _educated_ response.
> 
> If you don't want others to change your code just write it in the readme file, or better, make the repo private so that nobody could see it. What exactly are you trying to do?
> 
> Also, some minor improvements to your sentence. I'm pretty sure that this sentence won't pass the [HSK](https://en.wikipedia.org/wiki/Hanyu_Shuiping_Kaoshi) check.
> 
> ```diff
> - 看得出来你的语气有所悔改. 我会再次关闭该issue, 创建一个新的`clippy pass`的issue. 该次争吵希望你能认识到自己语气上的问题, 我宣告该次争论到此为止.
> + 看得出来你有所悔改。我会再次关闭该issue并创建一个新的`clippy pass` issue。希望你能认识到自己在该次争吵中语气上的问题，我宣告该次争论到此为止。
> ```

I think many cannot pass HSK. I have seen once HSK in my school. School classmates told me that the Chinese grammar is usually be ignored for local. The is not important. However, I do not understand why he use English punctuation in Chinese sentences.

**Comment by lisonge** [comment]
话说你们学 rust 的高中生一个个咋这么强

我在高中的时候还在被窝里打胶

**Comment by Bylx666** [comment]
十分感谢各位对我的耐心批评和指导. 经过一段时间反思, 我理解并决定承认了我的两点错误: 

1. 我的仓库的issue区域属于我的客厅, 无论客人态度如何, 我不该意气用事以牙还牙. 不论其参考价值如何, 我不该发出自己没有完全把握的发言. 
2. 我的作品不论是否有人来读源码, 我应当负起对其的责任心, 至少应当整理好她的基本仪容仪表. 我至少应当将我的代码整理得清晰顺畅, 暂且不论实现如何, 也不该将一份完全没有经过自我审视的代码开源并过大宣传. 

clippy和fmt这些应当只是我的代码实现的审查底线, 而不是我作品的优化目标之一. 我承认我的这些过错, 并对我伤害过的人道歉. 

起初我认为这只是一场无聊的网暴, 但大家都十分耐心并详细地指出了我的问题并给了我不少选择. 我十分感激各位的耐心, 希望大家不会厌恶看到一位少年的成长. 

我并没有想到我送给自己的成人礼会以这样的形式由我自己查收, 但我认为这是一次不错的经历. 我对大家的态度由一开始的费解和感到无聊, 到现在的理解和感激, 我想一定少不了大家对我成长的期待. 

最后再次感谢大家的耐心!

**Comment by hadesxiong** [comment]
后生可畏，激励我一个大龄社畜也要好好学习。

**Comment by rookie-luochao** [comment]
不管争论结果怎么样，后生可畏，都只是高中，大龄码畜也要继续好好学习了

**Comment by supermanmsc** [comment]
> 一丝不挂地表达出自己的优越, 我认可你的勇气. 这说到底是我给自己写的东西, 一开始就没抱多大期望有很多人给我改代码.
> 
> 说到底你所谓的"优化"或许只是你的一厢情愿罢了. 用Rust贴近汇编的思维方式并不可耻, 懒得写生命周期, "滥用"static mut同样是我自己的选择.
> 
> 我不会关闭该issue, 我也不反对你说的问题. 这一切是我咎由自取, 我会将此issue挂起, 在有所谓的"优化"的准备时拿来参考.
> 
> 十分感谢你的直言不讳.

我支持作者,  开源是作者的选择, 代码如何写也是个人选择, 跟任何人无关,  我也觉得你这个评级很傻x

![image](https://github.com/Bylx666/key-lang/assets/12078448/c15bdfb2-c153-48fd-8878-33be9f0017ef)

你会用rustfmt, 会用 vscode 一定很牛吧,  可这些东西对一个写编译器的人, 会不屑一顾,  
先辈们为了在arm上快速的做除法, 也会搞一些莫名其妙的数字.
加解密库里, 这种幻数更是很多. 这能说明什么呢?  说明他们不会vscode吧.....

搞计算机的人, 都应该是个侠客吧, 也会练武术,  
会用vscode只是会点花招, 
会算法数据结构, 也算会内功,
会编译原理,系统内核,硬件原理,也算会心法.

不在一个层次的人, 没法交流
提议作者不要理他.

**Comment by WenSimEHRP** [comment]
> 不在一个层次的人, 没法交流
> 提议作者不要理他.

I guess you are bringing up another fight. How's that good for you or the author? You are not going to take any responsibility for that, right? Why would you convince the author to do so even if they had already apologized?

> 我支持作者, 开源是作者的选择, 代码如何写也是个人选择, 跟任何人无关, 我也觉得你这个评级很傻x

Look, if a piece of code couldn't even meet the barebones of a language... Then what is the point of sharing it online? Just to show off and tell the community "look I have a compiler!" And ashame themselves?

You wrote "...how the code is written is also a personal choice." I just... Cannot really understand what you mean by that. The PR is about code formatting, not changing the algorithm or stuff. If it is the latter then of course the author could say "it is a personal choice" and reject the changes. But how a piece of code is formatted determines how other people could understand the code, and it is even related to its stability. True, in the license it says "absolutely no warrenty" but putting such things online is not taking the responsibility of their own work, and is also being irresponsible to the community.

I suggest you to go over the whole conflict first and then provide suggestions. If you have already done that I have nothing to say then.

> 先辈们为了在arm上快速的做除法, 也会搞一些莫名其妙的数字.
加解密库里, 这种幻数更是很多. 这能说明什么呢? 说明他们不会vscode吧.....

How is that related to the question? Those are just random examples and they are absolutely off-topic. How's the problem related to vscode?

||*And why are you using English punctuations? Is that a personal choice or you are taught to do so?*||

**Comment by else05** [comment]
字太多了，没细看，无脑的说一句，我还是鼓励一下，作为一个高中生已经站在线上准备起跑了，并不是坏事，成长是一个过程，老程序员也没必要拿着多年的工作积累和一个刚起步的人对线，不对等的，好比说是18岁了和小你10岁的人比身高，

**Comment by aurexav** [comment]
谁还没有一个 17 岁呢？

说实话，看起来年龄都不大，既然是学做长辈的样子，那么也请学着包容一些。
每个人的 GitHub Profile 都看了一遍，起码仓库作者确实做了些东西放了出来。

Rust 这门语言真的邪门，高傲从核心开发组开始不停地向下传染。

---

一码归一码，有错就接受吧。
聚焦在问题上，而不是聚焦在解决提出问题的人身上。

---

愿意花心思去阅读代码提出 issue，也不要相爱相杀了最后。

**Comment by ice1000** [comment]
感觉黑的过头了，用这些特性或者不开 clippy 也没什么不好的。

**Comment by PhotonQuantum** [comment]
事实上原实现是可以通过 miri 检查的，虽然存在潜在的内存泄漏，但是[内存泄漏是内存安全的](https://doc.rust-lang.org/book/ch15-06-reference-cycles.html)

```rust
static mut GLOBAL_STR: &str = "Hello, world!";

fn main() {
    // Read from stdin
    let mut buf = String::new();
    std::io::stdin()
        .read_line(&mut buf)
        .expect("Failed to read from stdin");

    // Replace content in GLOBAL_STR with buf
    let mut old_str = unsafe { &mut GLOBAL_STR };
    println!("old_str: {}", old_str);
    *old_str = buf.leak();
    // Oops: old_str leaked!

    // Print new GLOBAL_STR.
    println!("GLOBAL_STR: {}", unsafe { GLOBAL_STR });
}
```

```shell
$ MIRIFLAGS="-Zmiri-disable-isolation" cargo +nightly miri run
```

当然这里还是建议用 UnsafeCell 包裹一下避免 static mut 的使用，个人认为 rust 的 static mut 设计本身就是个败笔
```rust
#![feature(sync_unsafe_cell)]

use std::cell::SyncUnsafeCell;

static GLOBAL_STR: SyncUnsafeCell<&str> = SyncUnsafeCell::new("Hello, world!");

unsafe impl<T> Sync for AssertSync<T> {}

fn main() {
    // read from stdin
    let mut buf = String::new();
    std::io::stdin()
        .read_line(&mut buf)
        .expect("Failed to read from stdin");

    // replace content in GLOBAL_STR with buf
    unsafe {
        // SAFETY: only one thread is running this code
        let global_str_mut = &mut *GLOBAL_STR.get();
        let old_str = std::mem::replace(global_str_mut, buf.leak());
        println!("old_str: {}", old_str);
    }
    // old_str leaked! although this is memory safe...

    // print GLOBAL_STR
    // SAFETY: only one exclusive reference to GLOBAL_STR created.
    println!("GLOBAL_STR: {}", unsafe { *GLOBAL_STR.get() });
}
```

同样可以通过 miri 检查

不过其实最好的实现方法是针对每次执行建立一个独立的执行环境结构体，可参考 rhai 的实现
https://github.com/rhaiscript/rhai/blob/390d55a6e78ecaad987c25bd70a5f3bdf161dcaa/src/engine.rs#L88

或者参考 miri 的实现，其程序运行状态整个被包裹在 `Machine` 中，例如本项目中的 `LINE`（应该是程序计数器 Program Counter？）在 miri 实现中的对应为 `Machine` -> `Thread` -> `Frame` -> `loc`
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_const_eval/interpret/struct.Frame.html#structfield.loc

这样就可以尽可能避免全局变量，同时采用结构体包裹虚拟机可以通过 `&mut self` 的方式独占引用，避免掉很多 rust 的生命周期及同步原语的麻烦

**Comment by liaoxuewei** [comment]
> 你都给我看笑了, 希望你在1年之后回来看你的的发言不会脸红

为什么那么多人反对作者？奇怪了！

**Comment by transtone** [comment]
![image](https://github.com/Bylx666/key-lang/assets/46688/c83db866-7f8e-4b59-8a90-6a1b57d40daa)

在紧张的高中，应对高考之余还能有闲暇编码回帖，已经超过大多数同龄人了。
千里之行始于足下。开创语言，迈入成年，请从尊重文字，尊重承诺开始吧。

**Comment by aurexav** [comment]
> 为什么那么多人反对作者？奇怪了！

都是小圈子，观察一下互相点赞的。

**Comment by yongrongwang** [comment]
> 代码如何写也是个人选择, 跟任何人无关

就这种态度，你不适合跟别人一起做项目，你的代码大概率没人愿意看

**Comment by NeroBlackstone** [comment]
😅rust果然编程语言原神

**Comment by lizyChy0329** [comment]
楼主怕不是小时候被驴子踢过下脑袋

**Comment by qinsg** [comment]
支持一下项目的作者。只能讲后生可畏啊。想想我自己17岁还没接触计算机呢。加油！

**Comment by dylech30th** [comment]
作为一个玩具性质的开源项目代码写的随便一点是什么不能容忍的问题吗？
如果真要讨论的话能不能集中在更有价值的话题上，而不是代码写的好不好看这么做以后招不招同事待见这种我不仔细看还以为是什么经典老程序员在这训斥我不遵守公司的 Java 规范呢

**Comment by shaokeyibb** [comment]
你要觉得人家写的不好你可以提 pr，没必要拉上来个 issue 搁这儿秀你的优越感，真以为自己比别人懂点就可以理直气壮的开喷？

**Comment by ghost** [comment]
> 你要觉得人家写的不好你可以提 pr，没必要拉上来个 issue 搁这儿秀你的优越感，真以为自己比别人懂点就可以理直气壮的开喷？

你了解事件经过吗？#12是在PR被拒绝了发出的

**Comment by ghost** [comment]
如果你也拉黑了这个账号，那这件事会显得很戏剧
![image](https://github.com/Bylx666/key-lang/assets/169153711/e371c37a-3d21-4597-9877-236d1584e2ef)

**Comment by ghost** [comment]
> 字太多了，没细看，无脑的说一句，我还是鼓励一下，作为一个高中生已经站在线上准备起跑了，并不是坏事，成长是一个过程，老程序员也没必要拿着多年的工作积累和一个刚起步的人对线，不对等的，好比说是18岁了和小你10岁的人比身高，

谢谢，我15

**Comment by xianquu** [comment]
就是毫无价值的 发明这个语言干什么？已经有很多语言了，而且还不遵循rust规范🌝但是 那个发布issue的人也是闲着没事干 不喜欢就不理他不就行了🌝

@twhice @Bylx666

**Comment by hughdazz** [comment]
issue作者虽然学会了rustfmt、cargo check、vscode等，却没有学会尊重。

**Comment by Decodetalkers** [comment]
只是人家的个人项目，你觉得有问题提pr就是，个人项目对项目风格有绝对的所有权，任何人没有资格评价啊，而且clippy也只是建议，也可以关掉clippy，如果觉得这种风格是必要的。你迟早会遇到必须关闭clippy的情况的

**Comment by zhicheng233** [comment]
你们说的对但该项目使用MPL 协议Bylx666不需要对该项目做任何保证
还有这就是rust圈嘛，畏惧了，~~咱还是乖乖地待在Java吧~~

**Comment by MIDE-589** [comment]
30年前，我在同样年轻的时候，也试着发明一种编程语言。当时除了C语言我没有别的选择，我承认我当时干得糟糕透了，一些极其简单的功能却要消耗我大量的时间，而当时甚至连互联网都没有。最后我不得不放弃。如今我依然没有开发一种编程语言的实力。
年轻真好，可以勇敢地做一些以后再也不敢去尝试的事，可以有大把的时间去学习，可以尽情嘲笑我们这些为柴米油盐发愁的老家伙们，我真心羡慕你们。
年老也真好，至少我可以有机会接触那些简单而朴实的软硬件环境，知道它们运作的底层机理——好吧，其实是不得不学。

**Comment by Jedsek** [comment]
> 就是毫无价值的 发明这个语言干什么？已经有很多语言了，而且还不遵循rust规范🌝但是 那个发布issue的人也是闲着没事干 不喜欢就不理他不就行了🌝
> 
> @twhice @Bylx666

就是毫无价值的 写代码干什么？已经有很多代码了，而且还不如别人写的🌝但是 这个发布评论的人也是闲着没事干 不喜欢就不理他们不就行了🌝

@xianquu

**Comment by zhicheng233** [comment]
我的评价是写着来给自己玩玩可以，不过原作者似乎确实过度宣传了 ，会误导新人，更何况该项目是MPL协议，只能说这次双方都很抽象
________________________________
发件人: Jedsek ***@***.***>
发送时间: 2024年5月7日 20:44
收件人: Bylx666/key-lang ***@***.***>
抄送: 志成zhi_cheng ***@***.***>; Comment ***@***.***>
主题: Re: [Bylx666/key-lang] 优化代码审查 (Issue #12)


就是毫无价值的 发明这个语言干什么？已经有很多语言了，而且还不遵循rust规范🌝但是 那个发布issue的人也是闲着没事干 不喜欢就不理他不就行了🌝

@twhice<https://github.com/twhice> @Bylx666<https://github.com/Bylx666>

就是毫无价值的 写代码干什么？已经有很多代码了，而且还不如别人写的🌝但是 这个发布评论的人也是闲着没事干 不喜欢就不理他们不就行了🌝

@xianquu<https://github.com/xianquu>

**Comment by TabNahida** [comment]
> 30年前，我在同样年轻的时候，也试着发明一种编程语言。当时除了C语言我没有别的选择，我承认我当时干得糟糕透了，一些极其简单的功能却要消耗我大量的时间，而当时甚至连互联网都没有。最后我不得不放弃。如今我依然没有开发一种编程语言的实力。 年轻真好，可以勇敢地做一些以后再也不敢去尝试的事，可以有大把的时间去学习，可以尽情嘲笑我们这些为柴米油盐发愁的老家伙们，我真心羡慕你们。 年老也真好，至少我可以有机会接触那些简单而朴实的软硬件环境，知道它们运作的底层机理——好吧，其实是不得不学。

老前辈阿，现在方便多了有LLVM，或许你现在如果想的话可以尝试自己写一个语法解析器，然后用你熟悉的C语言做后端。现在的vlang就是这样做的。

**Comment by TabNahida** [comment]
> 😅rust果然编程语言原神

You're right, but Genshin Impact is a new open-world adventure developed by HoYoverse itself. The game takes place in a fantasy world called Teyvat, where those chosen by the gods are granted the "Vision" , the power of the guiding elements. You will play a mysterious character named "Traveler", in the free travel to meet different personalities, unique abilities of the companions, together with them to defeat the enemy, find lost relatives -- and gradually discover the truth of the "Allogene".

**Comment by zhangjh** [comment]
傻逼是真的多啊，满18岁了吗。。

**Comment by NaturalSelect** [comment]
说个实话，我不知道你在破防啥，既然你有open pull request的权利，repository的所有者也有权利拒绝它。

如果你确实需要你的commit，可以自己维护一个fork而不是来这里阴阳怪气。

**Comment by zhleyai** [comment]
这两个人一个十七，一个十五来着

--------------原始邮件--------------
发件人："zhangjh ***@***.***&gt;;
发送时间：2024年5月7日(星期二) 晚上10:32
收件人："Bylx666/key-lang" ***@***.***&gt;;
抄送："ZhleYai ***@***.***&gt;;"Comment ***@***.***&gt;;
主题：Re: [Bylx666/key-lang] 优化代码审查 (Issue #12)
-----------------------------------

 

 
傻逼是真的多啊，满18岁了吗。。

**Comment by rainzee** [comment]
捏着鼻子都觉得味儿大，自己代码没写过几行，优越感倒是强地不行，看了你的 repo 代码懒得看了，你张口闭口规范的，自己的 commit msg 写的什么乱七八糟的。可以进博物馆的蠢货。

**Comment by TabNahida** [comment]
> 谁还没有一个 17 岁呢？
> 
> 说实话，看起来年龄都不大，既然是学做长辈的样子，那么也请学着包容一些。 每个人的 GitHub Profile 都看了一遍，起码仓库作者确实做了些东西放了出来。
> 
> Rust 这门语言真的邪门，高傲从核心开发组开始不停地向下传染。
> 
> 一码归一码，有错就接受吧。 聚焦在问题上，而不是聚焦在解决提出问题的人身上。
> 
> 愿意花心思去阅读代码提出 issue，也不要相爱相杀了最后。

Rust是这样的

**Comment by ghost** [comment]
> 捏着鼻子都觉得味儿大，自己代码没写过几行，优越感倒是强地不行，看了你的 repo 代码懒得看了，你张口闭口规范的，自己的 commit msg 写的什么乱七八糟的。可以进博物馆的蠢货。

**自己代码没写过几行**？可以，我又成云coder了

**优越感倒是强地不行**你了解事件经过了吗？就这样就开始了你的评论？

我承认**抛开代码不谈**我的commit信息确实是乱写的，因为除了我没人看，我也没有去宣传它，我可从来没说过它们多么精致云云

**抛开代码不谈**，那么，我这个issue谈的不是代码说是什么？

**Comment by xiaodong2008** [comment]
**一些个人观点**
- 别人指出错误的时候应该虚心聆听，至于语气，如果你认为他说的是错的，你当然可以反驳，但是在你现在这个年纪，你真的能在这种情况下分辨出对错吗，没有一点意气用事的成分吗
- `省心, 最精致`，这两个关键词，明显和你的代码风格完全不同
- 代码的规范是重要的，拿我自己举例，除了一些临时的文件或者项目，都会注重可维护性，可读性，你既然能写出来 `省心, 最精致` 那就证明你不止是想自娱自乐，发到GitHub上也是想分享你的项目给别人
- 至于Issue的发起人是否有素质这件事：我不清楚事情的开始是怎样的，但即使一个人在言语上并不是那么的支持，但并不代表他提的意见一定不值得采纳，至少他有付出实际行动来支持你的项目

**带给项目作者的一段话**
非常不错，在这个年龄有自己的兴趣爱好写出这种项目值得鼓励，但不应该骄傲，保持虚心，向更高处前进

**自述**
不会Rust的菜鸡，一个写前端的

**Comment by yongrongwang** [comment]
我看了他们聊天记录，争吵是由不友好的提问者发起的，主要问题是语气很不好，但提到的问题无可厚非。正是语气不好让项目作者完全听不进去，但其实话糙理不糙，作者选择以牙还牙，好在最后作者虚心听取了建议。有人说项目作者17岁，但其实提问题的只有15岁，他俩对线叫年少气盛，吃瓜群众里面激动到骂人的年龄有10岁吗？这些人 low 到爆，还不如两个未成年成熟。

**Comment by TabNahida** [comment]
> > 捏着鼻子都觉得味儿大，自己代码没写过几行，优越感倒是强地不行，看了你的 repo 代码懒得看了，你张口闭口规范的，自己的 commit msg 写的什么乱七八糟的。可以进博物馆的蠢货。
> 
> **自己代码没写过几行**？可以，我又成云coder了
> 
> **优越感倒是强地不行**你了解事件经过了吗？就这样就开始了你的评论？
> 
> 我承认**抛开代码不谈**我的commit信息确实是乱写的，因为除了我没人看，我也没有去宣传它，我可从来没说过它们多么精致云云
> 
> **抛开代码不谈**，那么，我这个issue谈的不是代码说是什么？

Actually I never cared about my commit msg. In many projects, commit messages are not clear. It does not matter. Commit messages is just to remain you what did you commit. By the way, I think the README of Bylx666's profile is less readable.

**Comment by dylech30th** [comment]
> **一些个人观点**
> 
> * 别人指出错误的时候应该虚心聆听，至于语气，如果你认为他说的是错的，你当然可以反驳，但是在你现在这个年纪，你真的能在这种情况下分辨出对错吗，没有一点意气用事的成分吗
> * `省心, 最精致`，这两个关键词，明显和你的代码风格完全不同
> * 代码的规范是重要的，拿我自己举例，除了一些临时的文件或者项目，都会注重可维护性，可读性，你既然能写出来 `省心, 最精致` 那就证明你不止是想自娱自乐，发到GitHub上也是想分享你的项目给别人
> * 至于Issue的发起人是否有素质这件事：我不清楚事情的开始是怎样的，但即使一个人在言语上并不是那么的支持，但并不代表他提的意见一定不值得采纳，至少他有付出实际行动来支持你的项目
> 
> **带给项目作者的一段话** 非常不错，在这个年龄有自己的兴趣爱好写出这种项目值得鼓励，但不应该骄傲，保持虚心，向更高处前进
> 
> **自述** 不会Rust的菜鸡，一个写前端的

这不是纯纯拉偏架吗，库作者要改，issue 作者这个态度就没问题，双标的有点太明显了吧

**Comment by xiaodong2008** [comment]
我这里仅表达对于这个项目和技术有关的，不上升到素质等层面，这些太过于主观了，不好意思On 7 May 2024, at 23:39, Dylech30th ***@***.***> wrote:﻿

一些个人观点

别人指出错误的时候应该虚心聆听，至于语气，如果你认为他说的是错的，你当然可以反驳，但是在你现在这个年纪，你真的能在这种情况下分辨出对错吗，没有一点意气用事的成分吗
省心, 最精致，这两个关键词，明显和你的代码风格完全不同
代码的规范是重要的，拿我自己举例，除了一些临时的文件或者项目，都会注重可维护性，可读性，你既然能写出来 省心, 最精致 那就证明你不止是想自娱自乐，发到GitHub上也是想分享你的项目给别人
至于Issue的发起人是否有素质这件事：我不清楚事情的开始是怎样的，但即使一个人在言语上并不是那么的支持，但并不代表他提的意见一定不值得采纳，至少他有付出实际行动来支持你的项目

带给项目作者的一段话 非常不错，在这个年龄有自己的兴趣爱好写出这种项目值得鼓励，但不应该骄傲，保持虚心，向更高处前进
自述 不会Rust的菜鸡，一个写前端的

这不是纯纯拉偏架吗，库作者要改，issue 作者这个态度就没问题，双标的有点太明显了吧

**Comment by WenSimEHRP** [comment]
Github says 404...

DY_XiaoDong ***@***.***> 于 2024年5月7日周二 08:54写道：

**Comment by WenSimEHRP** [comment]
I guess the mailing function still works.

WenSim ***@***.***> 于 2024年5月8日周三 21:02写道：

**Comment by xiaodong2008** [comment]
Yes, the mailing function still works.

**Comment by LKBaka** [comment]
> 你都给我看笑了, 希望你在1年之后回来看你的的发言不会脸红

还是(孙)笑(川)大佬
